massa-factory-worker/src/denunciation_factory.rs

```diff
@@ -0,0 +1,260 @@
+                    if let Err(e) = de_ {
+                        debug!("Denunciation factory cannot create denunciation from block headers: {}", e);
+                        return;
+                    }
+                    Some(de_.unwrap()) // Already checked for error
+                } else {
+                    // Already 2 entries - so a Denunciation has already been created
+                    None
+                }
+            }
+            Entry::Vacant(ev) => {
+                ev.insert(vec![secured_header]);
+                None
+            }
+        };
+
+        if let Some(denunciation) = denunciation_ {
+            info!(
+                "Created a new block header denunciation : {:?}",
+                denunciation
+            );
+
+            let mut de_storage = self.channels.storage.clone_without_refs();
+            de_storage.store_denunciation(denunciation);
+            self.channels.pool.add_denunciations(de_storage);
+        }
+
+        self.cleanup_cache();
+    }
+
+    /// Process new secure share endorsement (~ endorsement)
+    fn process_new_secure_share_endorsement(
+        &mut self,
+        secure_share_endorsement: SecureShareEndorsement,
+    ) {
+        let key = (
+            secure_share_endorsement.content.slot,
+            secure_share_endorsement.content.index,
+        );
+        if self.endorsements_by_slot_index.contains_key(&key) {
+            warn!(
+                "Denunciation factory process an endorsement that have already been denounced: {}",
+                secure_share_endorsement
+            );
+            return;
+        }
+
+        if self.endorsements_by_slot_index.len() > DENUNCIATION_FACTORY_ENDORSEMENT_CACHE_MAX_LEN {
+            warn!(
+                "Denunciation factory cannot process - cache full: {}",
+                self.endorsements_by_slot_index.len()
+            );
+            return;
+        }
+
+        let denunciation_: Option<Denunciation> = match self.endorsements_by_slot_index.entry(key) {
+            Entry::Occupied(mut eo) => {
+                let secure_share_endorsements = eo.get_mut();
+                // Store at max 2 endo
+                if secure_share_endorsements.len() == 1 {
+                    secure_share_endorsements.push(secure_share_endorsement);
+
+                    // TODO / OPTIM?: use [] ?
+                    // safe to unwrap as we just checked the vec len
+                    let header_1 = secure_share_endorsements.get(0).unwrap();
+                    let header_2 = secure_share_endorsements.get(1).unwrap();
+                    let de_ = Denunciation::try_from((header_1, header_2));
+                    if let Err(e) = de_ {
+                        debug!("Denunciation factory cannot create denunciation from block headers: {}", e);
+                        return;
+                    }
+                    Some(de_.unwrap()) // Already checked for error
```

Comment:
use match instead of if Err + unwrap() ?

Reply:
fixed